src/apps/components/juttle-editor.js

```diff
@@ -0,0 +1,67 @@
+import React from 'react';
+
+import AceEditor from 'juttle-react-ace';
+
+import { promulgateBundle } from '../actions';
+import { LOCAL_BUNDLE_ID } from '../constants';
+
+import 'brace/theme/monokai';
+import './juttle-mode/mode';
+
+const FONT_SIZE = 14;
+const EDIT_EVAL_INTERVAL_MS = 1000;
+const SCROLL_MARGIN_TOP = 15;
+const SCROLL_MARGIN_BOTTOM = 15;
+
+class JuttleEditor extends React.Component {
+    onChange(newValue) {
+        if (!this.props.readOnly) {
+            this._updateProgram(newValue);
+            localStorage.setItem('program', newValue);
+        }
+    }
+
+    _updateProgram(newValue) {
+        clearTimeout(this.timeout);
+        this.timeout = setTimeout(() => {
+            let {juttleServiceHost, dispatch} = this.props;
+            let bundle = {program: newValue};
+            promulgateBundle(bundle, LOCAL_BUNDLE_ID, juttleServiceHost, dispatch);
+        }, EDIT_EVAL_INTERVAL_MS);
+    }
+
+    componentDidMount() {
+        let { editor } = this.refs.aceEditor;
+        editor.renderer.setScrollMargin(SCROLL_MARGIN_TOP, SCROLL_MARGIN_BOTTOM);
+        editor.setOptions({
+            maxLines: Infinity
+        });
+    }
+
+    render() {
+        let program = this.props.bundle ? this.props.bundle.program : '';
```

Comment:
the fact that we set the program to the bundle, which could potentially be a second old seems pretty rickety. Plus the `<AceEditor />` already keeps track of value state so now your syncing state in two different places and hoping things work out (This is causing the cursor skipping around error that I still experience from time to time). 

What I think we should do is set the value of the editor on `componentDidMount`: `editor.setValue(this.props.bundle)`.

Then, slap a `key` on the `JuttleViewer` class instance ie `<JuttleViewer key={bundleId} />`. This will ensure that whenever the bundle changes, `componentDidMount` will get called and we're properly setting the value of editor to the current bundle/program.